lms/djangoapps/discussion/rest_api/tests/test_render.py

```diff
@@ -56,33 +56,38 @@ def test_allowed_a_tag(self, protocol):
 
     def test_disallowed_a_tag(self):
         raw_body = '<a href="gopher://foo">link content</a>'
-        assert render_body(raw_body) == '<p>link content</p>'
+        assert render_body(raw_body) == '<p><a>link content</a></p>'
 
     @ddt.data("http", "https")
     def test_allowed_img_tag(self, protocol):
         raw_body = '<img src="{protocol}://foo" width="111" height="222" alt="bar" title="baz">'.format(
             protocol=protocol
         )
-        assert render_body(raw_body) == _add_p_tags(raw_body)
-
-    def test_disallowed_img_tag(self):
-        raw_body = '<img src="gopher://foo">'
-        assert render_body(raw_body) == '<p></p>'
+        self.assertHTMLEqual(render_body(raw_body), _add_p_tags(raw_body))
 
     def test_script_tag(self):
         raw_body = '<script type="text/javascript">alert("evil script");</script>'
         assert render_body(raw_body) == 'alert("evil script");'
 
-    @ddt.data("p", "br", "li", "hr")  # img is tested above
-    def test_allowed_unpaired_tags(self, tag):
+    @ddt.data(
+        ("br", '<p>foo<br>bar</p>'),  # br is allowed inside p
+        ("li", '<p>foo</p><li>bar<p></p></li>'),  # unpaired li only allowed if followed by another li
+        ("hr", '<p>foo</p><hr>bar<p></p>'),  # hr is not allowed inside p
+        ("img", '<p>foo<img>bar</p>'),  # unpaired img allowed, empty img doesn't render
+    )
+    @ddt.unpack
+    def test_unpaired_tags(self, tag, rendered_output):
         raw_body = f"foo<{tag}>bar"
-        assert render_body(raw_body) == _add_p_tags(raw_body)
+        assert render_body(raw_body) == rendered_output
 
     def test_unpaired_start_tag(self):
-        assert render_body('foo<i>bar') == '<p>foobar</p>'
+        assert render_body('foo<i>bar') == '<p>foo<i>bar</i></p>'
```

Comment:
this is applicable for all tags or only the "i" tag?

Reply:
Yes. I was focussed on getting the tests passing with as few changes as possible so it's easier to understand what's changed here.

Reply:
I'll just remove this test since this behaviour is no longer special.

Reply:
@xitij2000, let's add `'foo<i>bar' == '<p>foobar</p>'` and `'foo</i>bar' == '<p>foobar</p>'` to `test_unpaired_tags`, to be clear what's happening there.